src/display/editor/tools.js

```diff
@@ -59,6 +60,127 @@ class IdManager {
   }
 }
 
+/**
+ * Class to manage the images used by the editors.
+ * The main idea is to try to minimize the memory used by the images.
+ * The images are cached and reused when possible
+ * We use a refCounter to know when an image is not used anymore but we need to
+ * be able to restore an image after a remove+undo, so we keep a file reference
+ * or an url one.
+ */
+class ImageManager {
+  #baseId = getUuid();
+
+  #id = 0;
+
+  #cache = null;
+
+  async getFromFile(file) {
+    const { lastModified, name, size, type } = file;
+    const key = `${lastModified}_${name}_${size}_${type}`;
+    this.#cache ||= new Map();
+    let data = this.#cache.get(key);
+    if (data === null) {
+      return data;
+    }
+    if (data?.bitmap) {
+      data.refCounter += 1;
+      return data;
+    }
+
+    try {
+      data ||= {
+        bitmap: null,
+        id: `image_${this.#baseId}_${this.#id++}`,
+        file,
+        refCounter: 0,
+      };
+      data.bitmap = await createImageBitmap(file);
+      data.refCounter = 1;
+    } catch (e) {
+      console.error(e);
+      data = null;
+    }
+    this.#cache.set(key, data);
+    if (data) {
+      this.#cache.set(data.id, data);
+    }
+    return data;
+  }
+
+  async getFromUrl(url) {
+    this.#cache ||= new Map();
+    const key = url;
+    let data = this.#cache.get(key);
+    if (data === null) {
+      return data;
+    }
+    if (data?.bitmap) {
+      data.refCounter += 1;
+      return data;
+    }
+
+    try {
+      const response = await fetch(url);
+      const blob = await response.blob();
+      data ||= {
+        bitmap: null,
+        id: `image_${this.#baseId}_${this.#id++}`,
+        url,
+        refCounter: 0,
+      };
+      data.bitmap = await createImageBitmap(blob);
+      data.refCounter = 1;
+    } catch (e) {
+      console.error(e);
+      data = null;
+    }
+    this.#cache.set(key, data);
+    if (data) {
+      this.#cache.set(data.id, data);
+    }
+    return data;
+  }
```

Comment:
There appears to be fair amount of duplication in these methods, can we perhaps use a helper method to avoid that?
```suggestion
  async #get(key, rawData) {
    this.#cache ||= new Map();
    let data = this.#cache.get(key);
    if (data === null) {
      return data;
    }
    if (data?.bitmap) {
      data.refCounter += 1;
      return data;
    }

    try {
      let image;
      if (typeof rawData === "string") {
        const response = await fetch(rawData);
        image = await response.blob();
      } else {
        image = rawData;
      }
      data ||= {
        bitmap: null,
        id: `image_${this.#baseId}_${this.#id++}`,
        file,
        refCounter: 0,
      };
      data.bitmap = await createImageBitmap(file);
      data.refCounter = 1;
    } catch (e) {
      console.error(e);
      data = null;
    }
    this.#cache.set(key, data);
    if (data) {
      this.#cache.set(data.id, data);
    }
    return data;
  }
  
  async getFromFile(file) {
    const { lastModified, name, size, type } = file;
    return this.#get(`${lastModified}_${name}_${size}_${type}`, file);
  }

  async getFromUrl(url) {
    return this.#get(url, url);
  }
```